### PR TITLE
csi: add code for new CSI operator CR cephcluster

### LINKED ISSUE #14260: Integrate Rook with ceph-csi-operator
We are working on a cephcsi operator [1] which manages the life cycle of the cephcsi driver and also provides new CR's for easy installation and configuration of cephcsi drivers.

[1] https://github.com/ceph/ceph-csi-operator/

### PR DESCRIPTION
**Pull request by subhamkrai**
adding code changes,rbac changes required for create the new Ceph-CSI operator CR named cephCluster in api group 'csi.ceph.io'.

**Issue resolved by this Pull Request:**
Resolves #14260

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/r.yaml [hunk 1]
```diff
@@ -0,0 +1,418 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (subhamkrai) [inline deploy/examples/r.yaml:1]: this file will removed later as it added in this commit for my local testing purpose only

File: pkg/operator/ceph/csi/csi-operator.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/*
```
[the remark this answers is not in the document]
Comment (subhamkrai) [reply pkg/operator/ceph/csi/csi-operator.go]: agreed, done

File: pkg/operator/ceph/csi/operator.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+	csiCephCluster.Name = clusterName
+	csiCephCluster.Namespace = csiClusterConfig.Namespace
+	mons := sets.New[string]()
+	for _, mon := range csiClusterConfig.Monitors {
+		mons.Insert(mon.Endpoint)
+	}
+	csiCephCluster.Spec.Monitors = sets.List(mons)
+
+	err := r.client.Create(r.opManagerContext, csiCephCluster)
```
Comment (travisn) [inline pkg/operator/ceph/csi/operator.go:38]: If it already exists, we need update it (probably still just working on it while in draft).

File: pkg/operator/ceph/cluster/cluster.go [hunk 4]
```diff
@@ -230,12 +234,84 @@ func (c *ClusterController) initializeCluster(cluster *cluster) error {
...
+	if err != nil {
+		return err
+	}
+
+	if len(cephRBDMirrorList.Items) == 0 {
+		logger.Debug("no ceph subvolumegroup found")
+	} else {
+		if cephRBDMirrorList.Items[0].DeletionTimestamp.IsZero() {
+			createCSIOpCephClusterSpec.RbdMirrorDaemonCount = cephRBDMirrorList.Items[0].Spec.Count
+		}
```
Comment (subhamkrai) [inline pkg/operator/ceph/cluster/cluster.go:303]: @travisn is this right place to list and update the CSI-operator cephCLuster cr field `RbdMirrorDaemonCount`?

File: pkg/operator/ceph/cluster/cluster.go [hunk 5]
```diff
@@ -230,12 +234,84 @@ func (c *ClusterController) initializeCluster(cluster *cluster) error {
 	cluster.mons.ClusterInfo = cluster.ClusterInfo
 	cluster.mons.ClusterInfo.SetName(c.namespacedName.Name)
 
+	// Skip the new CSI-operator creation when holder pod is enabled until multus support is added in the CSI operator
+	if csi.EnableCSIOperator && !csi.IsHolderEnabled() {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/cluster.go:238]: In the mon reconcile there is code that updates the mons for the csi driver. I'd suggest we configure the CephConnections CR at that point. See [here](https://github.com/rook/rook/blob/54a294425274922b51abfb539059a4e7b248ec2c/pkg/operator/ceph/cluster/mon/mon.go#L1102-L1113) for internal clusters. See [here](https://github.com/rook/rook/blob/54a294425274922b51abfb539059a4e7b248ec2c/pkg/operator/ceph/cluster/cluster_external.go#L126-L139) for external clusters. We should probably refactor that to a single helper method though if we can avoid the duplication.

File: pkg/operator/ceph/cluster/cluster.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go:303]: @Madhu-1 We had some discussion about this rbd mirror daemon count, but I can't find the notes. What exactly is needed for that value? As I recall we don't actually need to query the cephrbdmirror CRs.

**Comment by subhamkrai** [comment]
closing this will continue here #14489

**Pull request closed without merging**